### DISCUSSION TITLE
Tasmota and "Public" MQTT server: no topics reported

### DISCUSSION THREAD

**Discussion by pdfkungfoo** [arendst/Tasmota#18654] [Q&A]
Because it will take a bit more time before I can start to setup and use my own private MQTT server, I tried to... 

- make my Tasmota plugs report to the "public" MQTT server at mqtt.eclipseprojects.io, and
- monitor my topics through the public MQTT web client at http://www.emqx.io/online-mqtt-client#/ (via mqttx.app/web)

After some experimentation with different settings (which all lead to Console reportings of `MQT: Connection refused`), I arrived at this local MQTT setup for the Tasmota plug:

<img width="290" alt="image" src="https://github.com/arendst/Tasmota/assets/4747960/488d7bf7-381f-4e54-91a2-a8cd3023df75">

...this setup on the MQTT server:

<img width="418" alt="image" src="https://github.com/arendst/Tasmota/assets/4747960/22cde22d-c01e-4f28-bc35-95dd711b7338">

...and this setup on the MQTT web client to look at the MQTT server:

<img width="826" alt="image" src="https://github.com/arendst/Tasmota/assets/4747960/0648dbc2-99b4-45de-90db-6514e69cecc9">

This seems to work, because clicking on the *"Connect"* link (top right corner of screenshot), confirms a successful connection (here from web MQTT client to MQTT server):

<img width="597" alt="image" src="https://github.com/arendst/Tasmota/assets/4747960/d53e186b-af44-4aab-9044-ce51f2559e78">

However, after creating a subscription to the topic *"kp_wh18_31415"*...

<img width="409" alt="image" src="https://github.com/arendst/Tasmota/assets/4747960/c733d691-2992-4d9e-b31b-515bc17a99d0">

...I see no MQTT messages arriving from the Tasmota plug.

Now I'm stuck at a point where Console reports these lines:

    00:00:00.001 HDW: ESP8266EX
    00:00:00.050 CFG: Loaded from flash at F8, Count 135
    00:00:00.055 QPC: Count 1
    00:00:00.073 NRG: Init driver 1
    00:00:00.076 Project tasmota - Tasmota-on-ShellyPlugS-001-PVModule Version 12.4.0(tasmota)-2_7_4_9(2023-02-24T15:32:38)
    00:00:00.566 WIF: Connecting to AP1 FRITZBox18 Channel 11 BSSId xx:xx:xx:xx:xx:xx in mode 11n as kp-wh18-31415-3350...
    00:00:01.756 WIF: Connected
    00:00:02.010 HTP: Web server active on kp-wh18-31415-3350 with IP address 192.168.180.72
    10:34:55.035 MQT: Attempting connection...
    10:34:55.454 MQT: Connected
    10:34:55.457 MQT: tele/kp_wh18_31415/LWT = Online (retained)
    10:34:55.459 MQT: cmnd/kp_wh18_31415/POWER = 
    10:34:55.464 MQT: tele/kp_wh18_31415/INFO1 = {"Info1":{"Module":"Shelly Plug S","Version":"12.4.0(tasmota)","FallbackTopic":"cmnd/mqttx_767c56b5/","GroupTopic":"cmnd/tasmotas/"}}
    10:34:55.474 MQT: tele/kp_wh18_31415/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"kp-wh18-31415-3350","IPAddress":"192.168.180.72"}}
    10:34:55.486 MQT: tele/kp_wh18_31415/INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":22}}
    10:34:55.497 MQT: stat/kp_wh18_31415/RESULT = {"POWER":"ON"}
    10:34:55.501 MQT: stat/kp_wh18_31415/POWER = ON
    10:34:58.411 QPC: Reset
    10:34:59.382 MQT: tele/kp_wh18_31415/STATE = {"Time":"2023-05-14T10:34:59","Uptime":"0T00:00:08","UptimeSec":8,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":" FRITZBox18","BSSId":"3C:A6:2F:4B:FC:3B","Channel":11,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":1,"Downtime":"0T00:00:03"}}
    10:34:59.414 MQT: tele/kp_wh18_31415/SENSOR = {"Time":"2023-05-14T10:34:59","ANALOG":{"Temperature":27.2},"ENERGY":{"TotalStartTime":"2023-03-03T10:42:21","Total":23.744,"Yesterday":0.309,"Today":0.076,"Period":0,"Power":37,"ApparentPower":52,"ReactivePower":37,"Factor":0.71,"Voltage":230,"Current":0.225},"TempUnit":"C"}

So the `MQT: Connected` message tells me that while Tasmota connected to the public MQTT server...
- ...I'm not seeing any submissions from Tasmota when searching (via the MQTT web client) the server for the topic I configured.

At this point I'm out of ideas what to try next.

**Comment by arendst** [comment]
You're probably almost there. Just configure the correct topic on the MQTT server: ``tele/kp_wh18_31415`` instead of ``kp_wh18_31415``

**Reply by sfromis** [reply]
If your topic has to start with your id, instead of the "prefix" (tele/stat/cmnd) as first level, you may want:
`FullTopic %topic%/%prefix%/`

**Reply by pdfkungfoo** [reply]
> [...] you may want: `FullTopic %topic%/%prefix%/`

While I don't yet have a grasp of that *%prefix%* stuff, shouldn't that rather be the other way round: `%prefix%/%topic%/` ?

**Reply by pdfkungfoo** [reply]
> You're probably almost there. Just configure the correct topic on the MQTT server: `tele/kp_wh18_31415` instead of `kp_wh18_31415`

I can't configure anything on the MQTT server, I can only configure things on the public web MQTT ***client*** (at http://www.emqx.io/online-mqtt-client#/).

If I use the `tele/` prefix for search/filter, I still don't see messages from my Tasmota plug.

I'll have to learn more about the MQTT protocol. Will search for good documents and do a round of reading up before continuing my experimentations.

Thanks for your suggestion.

**Reply by sfromis** [reply]
> shouldn't that rather be the other way round: `%prefix%/%topic%/`

That's the default, what you already have, where the "prefix" (tele/stat/cmnd) is at the start of the topic. If you have to have your id `kp_wh18_31415` at the start of the topic instead, you can change the topic order, as I mentioned. One case for this would be if the broker you connect to has a *requirement* for structure of the topics.

As a matter of taste, I've chosen to have my topic structure with the "prefix" not being at the start of the topic, but instead start with the name.

**Reply by fattaeld** [reply]
I hope you could solve this topic already. What I see on my side, this should work for you: "tele/kp_wh18_31415/SENSOR"
